bitbots_vision/bitbots_vision/yoeo_vision.py

```diff
@@ -0,0 +1,257 @@
+            self._vision_components.append(YOEOGoalpostDetectionComponent(self))
+        if new_config["component_line_detection_active"]:
+            self._vision_components.append(YOEOLineDetectionComponent(self))
+        if new_config["component_field_boundary_detection_active"]:
+            self._vision_components.append(YOEOFieldBoundaryDetectionComponent(self))
+        if new_config["component_field_detection_active"]:
+            self._vision_components.append(YOEOFieldDetectionComponent(self))
+        if new_config["component_debug_image_active"]:
+            self._vision_components.append(DebugImageComponent(self))
+
+        for vision_component in self._vision_components:
+            vision_component.configure(new_config, self._yoeo_handler)
+
+    def _register_subscribers(self, config: Dict) -> None:
+        self._sub_image = ros_utils.create_or_update_subscriber(
+            self,
+            self._config,
+            config,
+            self._sub_image,
+            'ROS_img_msg_topic',
+            Image,
+            callback=self._image_callback
+        )
+
+    def _image_callback(self, image_msg) -> None:
+        # TODO: type
+        """
+        This method is called by the Image-message subscriber.
+        Too old Image-Messages are dropped.
+
+        Sometimes the queue gets too large, even if the size is limited to 1.
+        That is why we drop old images manually.
+        """
+        logger.debug(f"{self.__class__.__name__}._image_callback(...) called")
+        if self._image_is_too_old(image_msg) or self._transfer_image_msg_mutex_is_locked():
+            return
+
+        with self._transfer_image_msg_mutex:
+            self._run_vision_pipeline_with(image_msg)
+
+    def _image_is_too_old(self, image_msg) -> bool:
+        image_age = self.get_clock().now() - rclpy.time.Time.from_msg(image_msg.header.stamp)
+        if 1.0 < image_age.nanoseconds / 1000000000 < 1000.0:
+            logger.warning(
+                f"Vision: Dropping incoming Image-Message because it is too old! ({image_age.to_sec()} sec)")
+            return True
+        else:
+            return False
+
+    def _transfer_image_msg_mutex_is_locked(self) -> bool:
+        return self._transfer_image_msg_mutex.locked()
+
+    @profile
+    def _run_vision_pipeline_with(self, image_msg):
+        """
+        Run the vision pipeline
+        """
+
+        image = self._convert_image_msg_to_cv2_image(image_msg)
+        if image is None:
+            logger.error("Vision pipeline - Image content is None")
+            return
+
+        self._yoeo_handler.set_image(image)
+        self._publish_image_to_components(image)
+
+        self._yoeo_handler.predict()
+        self._run_components(image_msg)
+
+    def _convert_image_msg_to_cv2_image(self, image_msg):
+        return self._cv_bridge.imgmsg_to_cv2(image_msg, 'bgr8')
+
+    def _publish_image_to_components(self, image) -> None:
```

Comment:
`Publish` is in the context of a pub/sub framework a bit misleading.
Instead, maybe:

- `_foreward_image_to_components`
- `_set_image_for_components`
- `_update_components_with_image`